src/Fixer/FunctionNotation/NoSpacesAfterFunctionNameFixer.php

```diff
@@ -61,40 +62,58 @@ public function isCandidate(Tokens $tokens)
      */
     protected function applyFix(\SplFileInfo $file, Tokens $tokens)
     {
-        $functionyTokens = $this->getFunctionyTokenKinds();
-        $languageConstructionTokens = $this->getLanguageConstructionTokenKinds();
-        $braceTypes = $this->getBraceAfterVariableKinds();
+        static $logicalTokenKinds = [
+            T_LOGICAL_AND,
+            T_LOGICAL_OR,
+            T_LOGICAL_XOR,
+        ];
+
+        $functionyTokenKinds = $this->getFunctionyTokenKinds();
+        $specialConstructTokenKinds = $this->getSpecialConstructTokenKinds();
+        $braceTokens = $this->getBraceAfterVariableTokens();
+
+        $tokensAnalyzer = new TokensAnalyzer($tokens);
 
         foreach ($tokens as $index => $token) {
-            // looking for start brace
+            // looking for start parenthesis
             if (!$token->equals('(')) {
                 continue;
             }
 
-            // last non-whitespace token, can never be `null` always at least PHP open tag before it
-            $lastTokenIndex = $tokens->getPrevNonWhitespace($index);
-
-            // check for ternary operator
-            $endParenthesisIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $index);
-            $nextNonWhiteSpace = $tokens->getNextMeaningfulToken($endParenthesisIndex);
-            if (
-                null !== $nextNonWhiteSpace
-                && $tokens[$nextNonWhiteSpace]->equals('?')
-                && $tokens[$lastTokenIndex]->isGivenKind($languageConstructionTokens)
-            ) {
-                continue;
+            // previous non-whitespace token, can never be `null` always at least PHP open tag before it
+            $prevNonWhitespace = $tokens->getPrevNonWhitespace($index);
+
+            // check for special construct with binary or ternary operator (or comma for echo)
+            if ($tokens[$prevNonWhitespace]->isGivenKind($specialConstructTokenKinds)) {
+                $endParenthesisIndex = $tokens->findBlockEnd(Tokens::BLOCK_TYPE_PARENTHESIS_BRACE, $index);
+                $nextMeaningful = $tokens->getNextMeaningfulToken($endParenthesisIndex);
+                if (
+                    null !== $nextMeaningful
+                    && (
+                        $tokens[$nextMeaningful]->isGivenKind(T_INSTANCEOF)
```

Comment:
Wondering: we are checking a list of potential tokens that might be found after the closing parenthesis that should prevent the space to be removed. Wouldn't it be easier to check whether the token *allows* to remove the space (i.e. `;`)?

Reply:
Hi, thanks for chiming in. I considered that (invert the existing logic), the list of tokens isn't much shorter:

  - `';'`, `T_CLOSE_TAG`
  - `')'`, `']'`, `'}'` plus the custom `CT::T_ARRAY_SQUARE_BRACE_CLOSE`, `CT::T_DYNAMIC_PROP_BRACE_CLOSE`, `CT::T_DYNAMIC_VAR_BRACE_CLOSE`, `CT::T_ARRAY_INDEX_CURLY_BRACE_CLOSE` (i.e. a "block end")
  - `':'`, `T_DOUBLE_ARROW`
  - `','` except for `echo`
  - `T_LOGICAL_AND`, `T_LOGICAL_OR`, `T_LOGICAL_XOR` for `print`
  - ... did I forget something?

That could give something like (can't test right now):

```php
                if (
                    null !== $nextMeaningful
                    && !(
                        $tokens[$nextMeaningful]->isGivenKind([T_CLOSE_TAG, T_DOUBLE_ARROW])
                        || $tokens[$nextMeaningful]->equalsAny([';', ':'])
                        || (!$tokens[$prevNonWhitespace]->isGivenKind(T_ECHO) && $tokens[$nextMeaningful]->equals(','))
                        || ($tokens[$prevNonWhitespace]->isGivenKind(T_PRINT) && $tokens[$nextMeaningful]->isGivenKind($logicalTokenKinds))
                        || (null !== ($blockType = Tokens::detectBlockType($tokens[$nextMeaningful])) && !$blockType['isStart'])
                    )
                ) {
                    continue;
                }
```

(and maybe "simplify" to

<details>

```php
                if (
                    null !== $nextMeaningful
                    && !$tokens[$nextMeaningful]->isGivenKind([T_CLOSE_TAG, T_DOUBLE_ARROW])
                    && !$tokens[$nextMeaningful]->equalsAny([';', ':'])
                    && ($tokens[$prevNonWhitespace]->isGivenKind(T_ECHO) || !$tokens[$nextMeaningful]->equals(','))
                    && (!$tokens[$prevNonWhitespace]->isGivenKind(T_PRINT) || !$tokens[$nextMeaningful]->isGivenKind($logicalTokenKinds))
                    && (null === ($blockType = Tokens::detectBlockType($tokens[$nextMeaningful])) || $blockType['isStart'])
                ) {
                    continue;
                }
```

</details>

) (and un-add `$tokensAnalyzer`), which could indeed handle more edge cases (for instance I didn't think to test `echo ($a ?: $b)[0];`, should try that when I can...)

What do you think?

Reply:
Well indeed, not really better... :)

Reply:
Wait, I have tested and all the following still have the space being removed:

`echo ($a ?: $b)->$c;`
`echo ($a ?: $b)::$c;`
`echo ($a ?: $b)($c);`
`echo ($a ?: $b)[$c];`
`echo ($a ?: $b){$c};`

So my current
```php
                        $tokens[$nextMeaningful]->isGivenKind(T_INSTANCEOF)
                        || $tokens[$nextMeaningful]->equals('?')
```
need to be extended to
```php
                        $tokens[$nextMeaningful]->isGivenKind([T_INSTANCEOF, T_OBJECT_OPERATOR, T_DOUBLE_COLON])
                        || $tokens[$nextMeaningful]->equalsAny(['?', '(', '[', '{'])
```
... The "inverted logic" seems more and more appealing...

Reply:
Heads-up: I have [first](https://github.com/FriendsOfPHP/PHP-CS-Fixer/compare/69e5d1ff3dddf9503b83cf6949367d873b51f41f..3bd7e0cd24d8cad9e1d58988ee1050ec23661afc) completed the tests and updated the code like in my last comment, and [finally](https://github.com/FriendsOfPHP/PHP-CS-Fixer/compare/3bd7e0cd24d8cad9e1d58988ee1050ec23661afc..dcca4fb65a77cbd357ba67b4d2c4c121c7c38cd5) inverted the logic like in my previous comment, which is simpler and more robust :slightly_smiling_face: